### PR TITLE
Prepare Vulkan layer in Orbit to use to run captures

### PR DESCRIPTION
**Pull request by kisenshi**
Use _SampleLayer_ as example to test the code and build a test layer in Orbit.

The working code has been updated, cleaned and organised. Also, the layer implementation corresponding to the test code has been removed. The current vulkan layer does not do much but it is ready to start implementing the logic related to running captures from it.

The steps followed have been:

- Update code to use absl for map and mutex
- Rename variables to follow cpp guidelines
- Clean code
- Isolate DispatchTable on its own class to make the code cleaner and more readable
- Remove test layer implementation calls and data

Ref: [renderdoc example](https://renderdoc.org/vulkan-layer-guide.html)

### CODE REVIEW

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (florian-kuebler) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:14]: Can you add the following packages to the bootstrap-orbit.sh (~line 29) and to the docker files (e.g. `third_party/conan/docker/Dockerfile.clang7 `) 
```
libvulkan-dev vulkan-validationlayers-dev
```

Comment (danielfenner) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:14]: Installing dev packages for dependencies locally on the build machines is somewhat of a last resort. We do this for qt to build the Linux UI. It is ok-ish to get this running experimentally but we should get this into conan/cmake since Vulkan is something we will make use of.
As I wrote in a chat to Cristina - I am mildly puzzeled why this even compiles on the CI. It seems as if the dev package are already installed accidentally.

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (florian-kuebler) [inline OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:59]: Again, please avoid abbreviations.

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (florian-kuebler) [inline OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:102]: Please rename.

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (florian-kuebler) [inline OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:137]: I am not sure about `strcpy` vs. `snprintf` @danielfenner or @ronaldfw maybe you do have some insights?

File: OrbitVulkanLayerClientGgp/include/OrbitVulkanLayerClientGgp/DispatchTable.h
Comment (florian-kuebler) [inline OrbitVulkanLayerClientGgp/include/OrbitVulkanLayerClientGgp/DispatchTable.h:16]: Please add a brief comment describing the purpose and the usage of this class.

**Review by florian-kuebler — COMMENTED** [review]
Thanks for that change. I left a couple of (mostly nit-picking) comments.

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (florian-kuebler) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:14]: >Installing dev packages for dependencies locally on the build machines is somewhat of a last resort. We do this for qt to build the Linux UI. It is ok-ish to get this running experimentally but we should get this into conan/cmake since Vulkan is something we will make use of.
As I wrote in a chat to Cristina - I am mildly puzzeled why this even compiles on the CI. It seems as if the dev package are already installed accidentally.

I discussed this with Henning a while ago, and we decided this will be the easiest and yet robust solution. This is also why the CI has these packages btw.

Comment (florian-kuebler) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:14]: But adding @beckerhe here.

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp:39]: The mutex are locked before the call. All the mutex logic is in the main file and transparent to this as it is handled from there. All calls are locked

Comment (florian-kuebler) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp:39]: I am not 100% sure about the scope of your layer. If it stays quite simple, this could be fine. Otherwise, I'd suggest to move it to this class (see the other comment).

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:33]: I considered moving it inside the class  when I cleaned the code but I realised the mutex was needed in the main file as well to read and process the data (e.g command buffer) so I decided to leave it in the main class as the same lock is used in both cases. I considered passing it as a parameter for example and lock inside the DispatchTable methods but I read that was not recommended so that's why I block it before the call.

Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:33]: I can update the code but I don't understand how I could lock the data safely in the main file if the mutex are different inside DispatchTable and there; they would not be locked safely in that case, would they?

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp:19]: You are right, I completely forgot to review the comments. I will go through all of them and fix/clarify or remove if necessary. Thanks!

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp:21]: Make sense, I will update this. Thanks!

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp:27]: I missed this too, makes sense; thanks!!

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp:50]: True, thanks for noticing!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (florian-kuebler) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:33]: Couldn't you use a different mutex for processing the data? Having one mutex for different data might only be relevant if there might be race conditions between the data.

Comment (florian-kuebler) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:33]: But as I said, it depends a bit of the size of your final layer. If it stays relatively small, it might be totally fine to have one mutex in the main.

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp]: Done!

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp]: Done!

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp]: Done!

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp]: Done!

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp]: I went through the comments and fixed/removed them. I left the one you mention below. Thanks!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: I have removed the comments in the method apart from this one, but I fixed it. Thanks!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: Done, thanks!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: Done, thanks!

File: OrbitVulkanLayerClientGgp/include/OrbitVulkanLayerClientGgp/DispatchTable.h
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/include/OrbitVulkanLayerClientGgp/DispatchTable.h:18]: Done, thanks!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:33]: I checked why in the tutorial they use a global mutex for both the table and the data and the reason they give is "We'll protect this by the same global lock as before to ensure multithreaded recording doesn't cause problems. " so I think it is needed to use the same lock.

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp]: This is being discussed in the other comment so I resolve this one.

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp]: This is being discussed in the other comment so I resolve this one.

File: OrbitVulkanLayerClientGgp/CMakeLists.txt
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/CMakeLists.txt:17]: Oh ok, will update the structure of the project then, thanks!

Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/CMakeLists.txt]: Done, thanks!

File: OrbitVulkanLayerClientGgp/CMakeLists.txt
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/CMakeLists.txt]: I don't understand what you mean but I have updated the CMakeLists.txt so please let me know if you think I should update something else. Thanks!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:40]: Oh, really? I was not aware of this; do you know why do they use it in the example tutorial?

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (florian-kuebler) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:33]: But you don't have any data to record so far, right? Anyways. For now it is also fine for me. Just wanted to check.

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (florian-kuebler) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:40]: Probably to have uniform code. The effect of the macro is really, just to have the symbol names as C style plain. So it would be really just be `OrbitCaptureClientCreateInstance`. This is required by all the functions, the loader needs to call directly, which are only the Get*ProcAddr once (as all others can be retrieved by those two).
So there is really no need in having this on all vulkan functions

**Review by florian-kuebler — APPROVED** [review]
Thanks for addressing the comments. LGTM.
Please wait until the remaining questions are resolved before merge.

File: OrbitVulkanLayerClientGgp/DispatchTable.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/DispatchTable.cpp]: Done, thanks!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: Oh I get it, I have updated the code, thanks!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: I did some investigation about this and it seems _snprintf_ is safer and recommended over strcpy so I have updated the code. Thanks!

Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: This seems to be breaking the build and it is related to the size; I am using `strlen `which should be returning the correct size? What should I do instead?

Comment (florian-kuebler) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: I am getting the same on my version. Haven't had the chance to look into this further. From my side, strcpy would be also fine. I'd be curious about  @danielfenner or @ronaldfw point of view.

Comment (ronaldfw) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: The second parameter can be 256 (the size of properties->layerName) here for this to be correct. The issue that you see is that strlen does not count the null termination character but that also needs to be copied. 

I would also suggest to use the Abseil version of snprintf: https://github.com/abseil/abseil-cpp/blob/e96d49687d9c078f2d47356b6723c3b5715493f7/absl/strings/str_format.h#L423

Comment (kisenshi) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: Updated, thanks!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (danielfenner) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp:14]: >I discussed this with Henning a while ago, and we decided this will be the easiest and yet robust solution. This is also why the CI has these packages btw.

Let's go with the dev-packages then for now. I'll chat with Henning to see if there is something to improve here.

**Review by danielfenner — APPROVED** [review]
lgtm!

File: OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp
Comment (ronaldfw) [reply OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp]: Latest version looks good to me!

**Review by ronaldfw — APPROVED** [review]
LGTM

**Comment by kisenshi** [comment]
Merging this now. Thanks!

**Pull request merged**
